versiondb/client/changeset.go

```diff
@@ -0,0 +1,322 @@
+	fp, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case strings.HasSuffix(fileName, ZlibFileSuffix):
+		zreader, err := zlib.NewReader(fp)
+		if err != nil {
+			_ = fp.Close()
+			return nil, err
+		}
+		reader = bufio.NewReader(zreader)
+	case strings.HasSuffix(fileName, SnappyFileSuffix):
+		reader = snappy.NewReader(fp)
+	default:
+		reader = bufio.NewReader(fp)
+	}
+	return WrapReader(reader, fp), nil
+}
+
+// withChangeSetFile opens change set file and pass the reader to callback,
+// it closes the file immediatelly after callback returns.
+func withChangeSetFile(fileName string, fn func(io.Reader) error) error {
+	reader, err := openChangeSetFile(fileName)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	return fn(reader)
+}
+
+// IterateChangeSets iterate the change set files,
+func IterateChangeSets(
+	reader io.Reader,
+	fn func(version int64, changeSet *iavl.ChangeSet) (bool, error),
+) (int64, error) {
+	var (
+		cont               = true
+		lastCompleteOffset int64
+		version, offset    int64
+		changeSet          *iavl.ChangeSet
+		err                error
+	)
+
+	for cont {
+		version, offset, changeSet, err = ReadChangeSet(reader, true)
+		if err != nil {
+			break
+		}
+		cont, err = fn(version, changeSet)
+		if err != nil {
+			break
+		}
+
+		lastCompleteOffset += offset
+	}
+
+	if err == io.EOF {
+		// it's not easy to distinguish normal EOF or unexpected EOF,
+		// there could be potential corrupted end of file and the err is a normal io.EOF here,
+		// user should verify the change set files in advance, using the verify command.
+		err = nil
+	}
+
+	return lastCompleteOffset, err
+}
+
+// IterateVersions iterate the version numbers in change set files, skipping the change set payloads.
+func IterateVersions(
+	reader io.Reader,
+	fn func(version int64) (bool, error),
+) (int64, error) {
+	var (
+		cont               = true
+		lastCompleteOffset int64
+		version, offset    int64
+		err                error
+	)
+
+	for cont {
+		version, offset, _, err = ReadChangeSet(reader, false)
+		if err != nil {
+			break
+		}
+		cont, err = fn(version)
+		if err != nil {
+			break
+		}
+
+		lastCompleteOffset += offset
+	}
+
+	if err == io.EOF {
+		// it's not easy to distinguish normal EOF or unexpected EOF,
+		// there could be potential corrupted end of file and the err is a normal io.EOF here,
+		// user should verify the change set files in advance, using the verify command.
+		err = nil
+	}
+
+	return lastCompleteOffset, err
+}
+
+// sortChangeSetFiles sort the change set files by the first version number in the file,
+// it also filters out the empty files,
+// assuming the file format is correct.
+func sortChangeSetFiles(files []string) ([]string, error) {
+	var (
+		nonEmptyFiles []string
+		versions      []uint64
+	)
+
+	for _, fileName := range files {
+		version, err := readVersion(fileName)
+		if err == io.EOF {
+			// skipping empty files
+			continue
+		}
```

Comment:
should handle other errors like failing to open the file?

Reply:
good catch, fixed.